### ISSUE TITLE
Callbacks in Meteor

### ISSUE THREAD

**Issue by aadamsx** [kriskowal/q#660]
I'm using Node.js calls on the server side in Meteor, outside of a Meteor-method -- specifically inside a server side Iron Router route.

A portion of the code inside of the route looks similar to the following so far:
```
  fs.mkdir(filebox, function (e) {
    if(!e || e.code === 'EEXIST') {
      fs.writeFile(file1, function (err) {
        if (err) throw err;

        fs.writeFile(file2, function (err) {
          if (err) throw err;

          fs.writeFile(file.3, function (err) {
            if (err) throw err;

            ExternalLibrary.do_something_(file1, function (err, buffer) {
              if (err) throw err;

              ExternalLibrary.do_something_(file2, function (err, buffer) {
                if (err) throw err;

                ExternalLibrary.do_something_(file3, function (err, buffer) {
                  if (err) throw err;

                  some_object = { first: file1, second: file2 }

                  ExternalLibrary.do_something_else_(some_object, function (err, buffer) {
                    if (err) throw err;

                    fs.readFile(file1, function (err, data) {
                      if (err) throw err;

                      res.write(data);
                      res.end();
                    });
                  });
                });
              });
            });
          });
        });
      });
    } else {
      console.log(e);
    }
  });
```
My problem is, I need to add even more calls to fs.write and the ExternalLibrary and further make these calls conditional.

It looks like I'm entering Callback hell.

I've read some on Node.js Promises and Generators. And specifically I'm trying out the frozeman/q-meteor library (which is just a wrapper on kriskowal/q).

What is the best way to write this using your Q library?

I have a SO post on this too: http://stackoverflow.com/questions/28886109/managing-asynchronous-callbacks-in-meteor

**Comment by kriskowal** [comment]
Callback hell can be mitigated with or without promises. The first step you should consider is to flatten out our chain with small named functions that forward the state that needs to be captured by future / inner callbacks explicitly, and discarding things in scope that do not need to be carried forward. It may be helpful to create a state object that can capture these values and pass them forward. An affordance of promises is that these small functions can simply be chained with `then`.

```js
var fs = require("q-io/fs");
return fs.makeDirectory(filebox)
.then(function () {
    return fs.writeFile(file1, content1);
}).then(function () {
    return fs.writeFile(file2, content2);
}).then(function () {
    return fs.writeFile(file3, content3);
}).done();
```

**Comment by kriskowal** [comment]
With generators, this is only slightly different.

```js
var fs = require("q-io/fs");
Q.async(function *() {
    yield fs.makeDirectory(filebox);
    yield fs.writeFile(file1, content1);
    yield fs.writeFile(file2, content2);
    yield fs.writeFile(file3, content3);
})().done();
```

And note that writing those files is probably parallelizable.

```js
return fs.makeDirectory(filebox)
.then(function () {
    return Q.all([
        fs.write(file1, content1),
        fs.write(file2, content2),
        fs.write(file3, content3)
    ]);
}).done();
```

And note that while parallelism is great, it can explode if you overwork a program, so if you've got enough files and you want to limit active parallelism, you might want to use a job queue of some sort. This is a feature of q-io@2, experimental.

```js
new Reader(files)
.forEach(function (file) {
    return fs.makeTree(fs.directory(file.path))
    .then(function () {
        return fs.write(file.path, file.content);
    })
}, this, 50).done();
```

**Comment by aadamsx** [comment]
Thanks a lot Kris.

I think I've almost got things working with your Q library.  One issue I can't get past is that some files can be written in parallel and sometimes I need the files to completely finish writing, as the next function needs access to them.

Here's what I have so far:
```
  makeDirectory(filebox)
    .then(writeFile(file1, data))
    .then(writeFile(file2, data))
    .then(writeFile(file3, data))
    .then(ExternalLibrary._DoSomethingWithWrittenFiles(file1))
    .then(ExternalLibrary._DoSomethingWithWrittenFiles(file2))
    .then(ExternalLibrary._DoSomethingWithWrittenFiles(file3))
    .then(ExternalLibrary._PerformAnotherOperationOnFiles(file1, file2))
    .then(readFile(filel1))
    .then(function (result) { //output from readFile
      console.log('after read file', result);
      res.write(result);
      res.end();
    })
    .catch(function (error) {
      console.log('catch error', error);
    })
    .done(function (result) {
      console.log('done', result);
    });
```

The writeFile methods can all run in parallel, but the ExternalLibrary functions need these files to finish writing completely before it can perform it's work -- *right now when node gets to the ExternalLibrary calls I get an error stating that the files don't exist*.  

This is what my deferred functions look like:
```
writeFile = function (dst, data) {
  console.log('desination:', dst);
  // console.log('data:', data);
  var fs = Npm.require('fs');
  var deferred = Q.defer();
  fs.writeFile(dst, data, function (err, result) {
    if (err) {
      deferred.reject(new Error(err));
    }
    else {
      deferred.resolve();
    }
  });
  return deferred.promise;
}

readFile = function (src) {
  var fs = Npm.require('fs');
  var deferred = Q.defer();
  fs.readFile(src, function (err, result) {
    if (err) {
      deferred.reject(new Error(err));
    }
    else {
      // console.log('readFile method, with result:', result);
      deferred.resolve(result);
    }
  });
  return deferred.promise;
}

makeDirectory = function (dst) {
  var fs = Npm.require('fs');
  var deferred = Q.defer();
  fs.mkdir(dst, function (err, result) {
    if(!err || err.code === 'EEXIST') {
      deferred.resolve();
    }
    else {
      deferred.reject(err);
    }
  });
  return deferred.promise;
}
```
And this is a sample of what the ExternalLibrary looks like:
```
 ExternalLibrary._DoSomethingWithWrittenFiles = function(file) {
   var deferred = Q.defer();
   ExternalLibrary.executeCommand([file, 'do_something', 'output ', output], function (error) {
     if (error) {
       console.log(error);
       deferred.reject(new Error(error));
     }
     else {
       deferred.resolve();
     }
   });
   return deferred.promise;
 };

ExternalLibrary.executeCommand = function (args, callback) {
  var command = 'file_command ' + args.join(' ');
  console.log(command);
  exec(command, {encoding: 'binary', maxBuffer: 1024 * 1000}, function(err, stdout, stderr) {
    if(err) return callback(new Error(err));
    callback(null);
  });
};
```
